spyder/widgets/editor.py

```diff
@@ -941,11 +947,73 @@ def __modified_readonly_title(self, title, is_modified, is_readonly):
             title = "(%s)" % title
         return title
 
-    def get_tab_text(self, filename, is_modified=None, is_readonly=None):
+    def get_tab_text(self, index, filename, is_modified=None, is_readonly=None):
         """Return tab title"""
-        return self.__modified_readonly_title(osp.basename(filename),
+        fname = self.get_file_title(index)
+        return self.__modified_readonly_title(fname,
                                               is_modified, is_readonly)
 
+    def get_file_title(self, index):
+        """Get tab title without ambiguation."""
+        finfo = self.data[index]
+        fname = osp.basename(finfo.filename)
+        for findex, file in enumerate(self.data):
+            fname_to_compare = osp.basename(file.filename)
+            if findex is not index and fname == fname_to_compare:
+                differ_path = self.differ_prefix(
+                                   self.path_components(finfo.filename), 
+                                   self.path_components(file.filename))
+                differ_path_length = len(differ_path)
+                if (differ_path_length > 20):
+                    path_components = self.path_components(differ_path)
+                    last_component_index = len(path_components) -1
+                    path_components = [path_components[0], '...', 
+                                       path_components[last_component_index]]
+                    differ_path = os.path.join(*path_components)
+                fname = fname + " - " + differ_path
+                break
+        return fname
+
+    def path_components(self, path):

```

Comment:
Let's move this function to `utils/sourcecode.py` and add a test for it. For that you need to move the tests at the end of that file to the `tests` folder in `utils`, in a file called `test_sourcecode.py`.